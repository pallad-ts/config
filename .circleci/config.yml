version: 2.1

orbs:
  node: circleci/node@4.7

jobs:
  main_job:
    executor:
      name: node/default
      tag: "14.18"
    steps:
      - checkout
      - run: "npm ci"
      - run:
          command: ./node_modules/.bin/lerna run compile
          name: compilation
      - run:
          command: ./node_modules/.bin/lerna run lint
          name: linter
      - run:
          command: ./node_modules/.bin/lerna run test
          name: test
      - store_test_results:
          path: ./test_results

workflows:
  main:
    jobs:
      - main_job
