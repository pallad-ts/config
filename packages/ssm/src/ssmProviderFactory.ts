import {SSM} from 'aws-sdk';
import {SSMProvider} from "./SSMProvider";
import DataLoader = require("dataloader");
import {wrapWithDefaultAndTransformer} from '@pallad/config';

const splitArray = require('split-array');

export function parameterDeserializer(parameter: SSM.Parameter) {
    if (parameter.Type === 'StringList') {
        return parameter.Value!.split(',');
    }
    return parameter.Value;
}

function createDataLoader(batchFN: ssmProviderFactory.BatchFunction) {
    return new DataLoader<string, SSMProvider.Value | undefined>(batchFN);
}

export function ssmProviderFactory<TTransformed, TDefault>(
    options?: ssmProviderFactory.Options
) {
    const ssm = options?.ssm ?? new SSM();
    const paramDeserializer = options?.parameterDeserializer ?? parameterDeserializer;

    const batchFn = async (keys: readonly string[]) => {
        const map = new Map<string, SSMProvider.Value | undefined>();
        const batches = splitArray(keys, 10);
        for (const batch of batches) {
            const parameters = await ssm.getParameters({
                Names: batch,
                WithDecryption: true
            }).promise();

            if (parameters.Parameters) {
                for (const parameter of parameters.Parameters) {
                    map.set(parameter.Name!, paramDeserializer(parameter));
                }
            }
        }
        return keys.map(x => map.get(x));
    };

    const dataLoader = (options?.createDataLoader ?? createDataLoader)(batchFn);
    return function <TTransformed, TDefault>(
        key: string,
        opts?: wrapWithDefaultAndTransformer.Options<TTransformed, TDefault, SSMProvider.Value>
    ) {
        const finalKey = (options?.prefix ?? '') + key;
        return wrapWithDefaultAndTransformer(
            new SSMProvider(finalKey, dataLoader),
            opts
        );
    };
}

export namespace ssmProviderFactory {
    export interface Options {
        ssm?: SSM;
        createDataLoader?: (batchFn: BatchFunction) => DataLoader<string, SSMProvider.Value | undefined>;
        parameterDeserializer?: ParameterDeserializer;
        prefix?: string;
    }

    export type BatchFunction = (key: readonly string[]) => Promise<Array<SSMProvider.Value | undefined>>;
    export type ParameterDeserializer = (parameter: SSM.Parameter) => Promise<any> | any;
}
