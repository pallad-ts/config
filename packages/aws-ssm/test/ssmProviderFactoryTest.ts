import { PutParameterCommand, PutParameterCommandInput, SSMClient } from "@aws-sdk/client-ssm";
import { ssmProviderFactory } from "@src/ssmProviderFactory";
import { left, right } from "@sweet-monads/either";
import { LocalstackContainer, StartedLocalStackContainer } from "@testcontainers/localstack";
// eslint-disable-next-line @typescript-eslint/naming-convention
import DataLoader from "dataloader";
import * as sinon from "sinon";



import { ValueNotAvailable } from "@pallad/config";


describe("ssmProviderFactory", () => {
    let container: StartedLocalStackContainer;
    let ssmClient: SSMClient;

    let helper: ReturnType<typeof ssmProviderFactory>;

    const CONFIG = {
        prefix: "pallad-config-",
    };

    async function createParameter(name: string, input: Pick<PutParameterCommandInput, "Value" | "Type">) {
        const result = await ssmClient.send(
            new PutParameterCommand({
                Name: CONFIG.prefix + name,
                ...input,
            })
        );
        console.log(result);
    }

    beforeAll(async () => {
        container = await new LocalstackContainer("localstack/localstack:3.6.0").start();
        ssmClient = new SSMClient({
            endpoint: container.getConnectionUri(),
            logger: console,
        });

        await Promise.all([
            createParameter("var1", { Value: "bar", Type: "String" }),
            createParameter("var2", { Value: "foo,bar", Type: "StringList" }),
            createParameter("var3", { Value: "world", Type: "String" }),
            createParameter("var4", { Value: "10", Type: "String" }),
            createParameter("var5", { Value: "wjeNdAbmyjtqAU2OcuPR15F4xDzlV1uZHmZNNl1p", Type: "String" }),
            createParameter("var6", { Value: "u6yjSXzp9q5yhaCdEWyj", Type: "String" }),
            createParameter("var7", { Value: "7FZxN9NHXcgleeP0TSva", Type: "String" }),
            createParameter("var8", { Value: "R3aogUWGdvDJy03brMVEZBnqKJYZcp", Type: "String" }),
            createParameter("var9", { Value: "XUlAtGjdCwIiHDnojsFqmOC6HO8SyXX", Type: "String" }),
            createParameter("var10", { Value: "u97tvnLBUOurP3FcpUWRqbO81GqqH", Type: "String" }),
            createParameter("var11", { Value: "HMZNkhmMOxZsylqOaxfNwjeAf9Ju", Type: "String" }),
            createParameter("var12", { Value: "DRJeSQn6CmLVuo5eXHGkVABTsgy", Type: "String" }),
        ]);
    }, 120000);

    beforeEach(() => {
        helper = ssmProviderFactory({
            ...CONFIG,
            ssm: ssmClient,
        });
    });

    afterAll(() => {
        return container.stop();
    });

    describe("loading", () => {
        it("success", () => {
            const provider = helper("var1");

            return expect(provider.getValue()).resolves.toEqual(right("bar"));
        });

        it("failure due to lack of parameter", () => {
            const dependency = helper("var1000");

            return expect(dependency.getValue()).resolves.toStrictEqual(
                left(new ValueNotAvailable("SSM: pallad-config-var1000"))
            );
        });

        it("loading array of strings", () => {
            const dependency = helper("var2");
            return expect(dependency.getValue()).resolves.toEqual(right(["foo", "bar"]));
        });

        it("uses custom dataLoader factory", () => {
            const spy = sinon.spy((batchFn: ssmProviderFactory.BatchFunction) => {
                return new DataLoader(batchFn);
            });

            ssmProviderFactory({
                ...CONFIG,
                createDataLoader: spy,
            });

            sinon.assert.calledOnce(spy);
        });

        it("fails if dataloader has misconfigured maxBatchSize", async () => {
            const provider = ssmProviderFactory({
                ...CONFIG,
                createDataLoader: (batchFn: ssmProviderFactory.BatchFunction) => {
                    return new DataLoader(batchFn);
                },
            });

            const result = await Promise.all(
                ["k1", "k2", "k3", "k4", "k5", "k6", "k7", "k8", "k8", "k10", "k11", "k12"].map(x => {
                    return provider(x).getValue();
                })
            );

            const finalResult = result.map(x => {
                return x.fold(
                    x => {
                        if (x instanceof Error) {
                            return x.message;
                        }
                    },
                    x => undefined
                );
            });

            expect(finalResult).toMatchSnapshot();
        });

        it("using parameter deserializer", () => {
            const stub = sinon.stub().resolves(1000);

            const helper = ssmProviderFactory({
                ...CONFIG,
                ssm: ssmClient,
                parameterDeserializer: stub,
            });

            const dependency = helper("var1");

            return expect(dependency.getValue()).resolves.toEqual(right(1000));
        });

        it("loading many variables", () => {
            const dependencies = [
                helper("var1"),
                helper("var2"),
                helper("var3"),
                helper("var4"),
                helper("var5"),
                helper("var6"),
                helper("var7"),
                helper("var8"),
                helper("var9"),
                helper("var10"),
                helper("var11"),
                helper("var12"),
            ];

            const resolvedDeps = Promise.all(dependencies.map(x => x.getValue()));
            return expect(resolvedDeps).resolves.toEqual(
                [
                    "bar",
                    ["foo", "bar"],
                    "world",
                    "10",
                    "wjeNdAbmyjtqAU2OcuPR15F4xDzlV1uZHmZNNl1p",
                    "u6yjSXzp9q5yhaCdEWyj",
                    "7FZxN9NHXcgleeP0TSva",
                    "R3aogUWGdvDJy03brMVEZBnqKJYZcp",
                    "XUlAtGjdCwIiHDnojsFqmOC6HO8SyXX",
                    "u97tvnLBUOurP3FcpUWRqbO81GqqH",
                    "HMZNkhmMOxZsylqOaxfNwjeAf9Ju",
                    "DRJeSQn6CmLVuo5eXHGkVABTsgy",
                ].map(right)
            );
        });
    });
});
