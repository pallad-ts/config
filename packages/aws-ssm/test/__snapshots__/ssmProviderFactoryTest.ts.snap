// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssmProviderFactory loading fails if dataloader has misconfigured maxBatchSize 1`] = `
[
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
  "Misconfigured dataloader. Missing required \`maxBatchSize: 10\` option",
]
`;
